src/mongocrypt.c

```diff
@@ -574,30 +530,403 @@ mongocrypt_init (mongocrypt_t *crypt)
-         if (candidate.okay) {
-            // We got one:
-            crypt->csfle_vtable = candidate.vtable;
-            crypt->csfle_lib = candidate.lib;
+         if (candidate_csfle.okay) {
             // Stop searching:
             break;
          }
       }
    }
+
    mstr_free (csfle_cand_filepath);
+   return candidate_csfle;
+}
+
+
+/// Global state for the application's csfle library
+typedef struct csfle_global_lib_state {
+   /// Synchronization around the reference count:
+   mongocrypt_mutex_t mtx;
+   int refcount;
+   /// The open library handle:
+   mcr_dll dll;
+   /// vtable for the APIs:
+   _mcr_csfle_v1_vtable vtable;
+   /// The global library state managed by the csfle library:
+   mongo_csfle_v1_lib *csfle_lib;
+} csfle_global_lib_state;
+
+csfle_global_lib_state g_csfle_state;
+
+static void
+init_csfle_state (void)
+{
+   _mongocrypt_mutex_init (&g_csfle_state.mtx);
+}
+mlib_once_flag g_csfle_init_flag = MLIB_ONCE_INITIALIZER;
+
+/**
+ * @brief Verify that `found` refers to the same library that is globally loaded
+ * for the application.
+ *
+ * @param crypt The requesting mongocrypt_t. Error information may be set
+ * through here.
+ * @param found The result of _try_load_csfle()
+ * @return true If `found` matches the global state
+ * @return false Otherwise
+ *
+ * @note This function assumes that the global csfle state is valid and will not
+ * be destroyed by any other thread. (One must hold the reference count >= 1)
+ */
+static bool
+_validate_csfle_singleton (mongocrypt_t *crypt, _loaded_csfle found)
+{
+   mongocrypt_status_t *status = crypt->status;
+   // Path to the existing loaded csfle:
+   mcr_dll_path_result existing_path_ = mcr_dll_path (g_csfle_state.dll);
+   assert (existing_path_.path.data &&
+           "Failed to get path to already-loaded csfle library");
+   mstr_view existing_path = existing_path_.path.view;
+   bool okay = true;
+   if (!found.okay) {
+      // There is one loaded, but we failed to find that same library. Error:
+      CLIENT_ERR ("An existing CSFLE libary is loaded by the application at "
+                  "[%s], but the current call to mongocrypt_init() failed to "
+                  "find that same library.",
+                  existing_path.data);
+      okay = false;
+   } else {
+      // Get the path to what we found:
+      mcr_dll_path_result found_path = mcr_dll_path (found.lib);
+      assert (found_path.path.data && "Failed to get the DLL path of the "
+                                      "library that we just barely found");
+      if (!mstr_eq (found_path.path.view, existing_path)) {
+         // Our find-result should only ever find the existing same library.
+         // Error:
+         CLIENT_ERR (
+            "An existing CSFLE library is loaded by the application "
+            "at [%s], but the current call to mongocrypt_init() attempted to "
+            "load a second CSFLE library from [%s]. This is not allowed.",
```

Comment:
It might be nice to have a comment in the source about the actual problem that this is solving.

Would the two library instances conflict with each other in some way? (The global state shouldn’t really conflict, esp. since we’re using `RTLD_LOCAL` on POSIX and Windows has that behavior anyway). At a glance, it seems like this would only be leaking memory, but that seems fine-ish.

Also, keep in mind that there may still be multiple copies of libmongocrypt loaded at the same time, or the same copy loaded and unloaded repeatedly, and consequently it’s hard to actually guarantee singleton-ness of the csfle shared library from here.

Reply:
Whether two loaded csfle libraries will conflict (even if they share distinct memory mappings) is up to the server team. The library they have given us assumes that it will be the only loaded instance for the entire process, and this code is just to ensure those guarantees as best as we can, even if not perfect.

Reply:
Thank you for the clarification – I only did a :+1: reaction and left this comment unresolved because I still think it might be nice to have a comment in the source about why we’re doing this, but ultimately, this concern is addressed :)

(It’s a bit sad that the server team can’t give us stronger guarantees here, but I also get why they don’t.)